# Call Center Cleanup

For this assignment, we will be working with call center data. You can start working on the assignment after the first lesson on Exploratory Data Analysis. Make sure to read the whole assignment before starting anything! As you code along in the Jupyter notebook, you are asked to make note of the results of your analysis. Do so by clicking on the results box and adding your notes beside each question.

## Business Issue and Understanding

You are working for a company that has two call centers: the North Call Center and the South Call Center. The company is looking to possibly hire five additional reps to enhance customer experience. Your task is to explore how efficient the current reps are in each branch to determine which branch would benefit from additional assistance.

### How the Call Center Works

Call center representatives are assigned queues. When calls are assigned to a queue, the call is assigned to the next person in line in the queue. After a call is assigned to a representative, the amount of time between assignment and the call starting is divided into busy minutes and not ready minutes. If the call is incoming and a customer is waiting on the phone for a rep, the time is split into three categories: busy minutes, not ready minutes, and incoming call wait time. Once the rep has the customer on the phone, there might be during call wait time, where the call is put on hold while the rep gets an answer for the customer.

### Notes about the Dataset

If you haven't worked in a call center before, these notes might help you throughout your analysis.

* The call purpose is tagged for each call.
* The time of the call is tagged in 1 hour blocks starting at 9:00 AM and ending at 5:00 PM.
* Calls are tagged as incoming or outgoing.
* Reps are assigned to queues. When the reps are working, they take calls in the order of their queue.
* A call that is dropped due to technical error or missed by the center because they have reached maximum capacity is a lost call.
* An abandoned call is when the customer hangs up because they have been waiting for too long.
* Busy Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are busy with other customers.
* Not Ready Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are not ready (for example, getting water).
* Incoming Wait Time - amount of time after assigned rep is available to take the call customer waits for representative to pick up a call. This is tracked in seconds.
* During Call Wait Time - amount of time during call that customer has to wait for representative

## Getting Started

You have two CSVs at your disposal, `NorthCallCenter.csv` and `SouthCallCenter.csv`. Import the appropriate libraries and create two dataframes, one called `north_df` and one called `south_df`.

In [5]:
# Import the appropriate libraries with aliases
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
# Create two new dataframes
north_df = pd.read_csv("NorthCallCenter.csv")
south_df = pd.read_csv("SouthCallCenter.csv")

<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Exploratory Data Analysis

Time to do some EDA! In the process of learning more about the two datasets, answer the following questions. In the code blocks below enter the code used to solve the question.  At the end of the section record your answers by double clicking where asked.

#### EDA Question 1A:  How many reps are in the North branch?  

In [47]:
#Code Used:
north_df.head()
countN = north_df['Rep ID'].drop_duplicates().size
print(countN)

9


#### EDA Question 1B:  How many reps are in the South branch?  

In [37]:
#Code Used:
south_df.head()
countS = south_df['Rep ID'].drop_duplicates().size
print(countS)

11


#### EDA Question 2A:  What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for the North branch? 

In [45]:
#Code Used:
north_df.head()
avg1 = north_df['Busy Minutes'].mean()
avg2 = north_df['Not Ready Minutes'].mean()
avg3 = north_df['Incoming Wait Time'].mean()
avg4 = north_df['During Call Wait Time'].mean()
print(avg1,avg2,avg3,avg4)

9.987755102040817 1.9142857142857144 3.048780487804878 2.9673469387755103


#### EDA Question 2B:  What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for the South branch? 

In [46]:
#Code Used:
avg5 = south_df['Busy Minutes'].mean()
avg6 = south_df['Not Ready Minutes'].mean()
avg7 = south_df['Incoming Wait Time'].mean()
avg8 = south_df['During Call Wait Time'].mean()
print(avg5,avg6,avg7,avg8)

10.054140127388536 1.910828025477707 3.0 3.0764331210191083


#### EDA Question 3:  What is the number of calls taken for each time block(Both North and South combined)?

In [18]:
# Code Used:
n_df = north_df.groupby(["Time Block"]).Calls.sum().reset_index()
s_df = south_df.groupby(["Time Block"]).Calls.sum().reset_index()
display(n_df)
display(s_df)

Time Block  Calls
0   10:00 AM     43
1   11:00 AM     25
2   12:00 PM     45
3    1:00 PM     17
4    2:00 PM     35
5    3:00 PM     30
6    4:00 PM     28
7    5:00 PM     17
8    9:00 AM      5

Time Block  Calls
0   10:00 AM     56
1   11:00 AM     31
2   12:00 PM     75
3    1:00 PM     23
4    2:00 PM     30
5    3:00 PM     43
6    4:00 PM     25
7    5:00 PM     26
8    9:00 AM      5

##### Record your answers for the EDA section by double clicking here:
- 1a) Reps in the North Branch = 9
- 1b) Reps in the South Branch = 11


- 2a) North Branch, (round to the nearest hundredth):
> - Busy Minutes = 9.99
> - Not Ready Minutes = 1.91
> - Incoming Wait Time = 3.05
> - During Call Wait Time = 2.97
- 2b) South Branch, (round to the nearest hundredth):
> - Busy Minutes = 10.05
> - Not Ready Minutes = 1.91
> - Incoming Wait Time = 3.00
> - During Call Wait Time = 3.08
- 3) Total Calls taken for all branches per time block:
> - 9:00AM   = 10
> - 10:00AM  = 99
> - 11:00AM  = 56
> - 12:00PM  = 120
> - 1:00PM   = 40
> - 2:00PM   = 65
> - 3:00PM   = 73
> - 4:00PM   = 53
> - 5:00PM   = 43



<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Cleaning Your Data 
Now you need to clean up the datasets. When cleaning the datasets, you may find that there isn't dirty data to clean. That is okay! Some questions you need to answer about the data sets.
* Add additional code blocks as needed to show how you came to your conclusions. Add comments in your code blocks to help others understand your thinking. 


#### Cleaning Question 1:  Is there any missing data and if yes explain what you would do with the missing data and why?
##### Record your answer by double clicking here:
> Your Answer: No missing data was found. The total amount of non-null count in both tables matches their numbers of total entries.

In [81]:
# Code Used
# Add Additional Cells as needed.
n2_df = pd.read_csv("NorthCallCenter.csv")
s2_df = pd.read_csv("SouthCallCenter.csv")
n2_df = n2_df.dropna()
n2_df.info()
s2_df = s2_df.dropna()
s2_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 82 entries, 0 to 81
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             82 non-null     int64  
 1   Branch                 82 non-null     object 
 2   Call Purpose           82 non-null     object 
 3   Time Block             82 non-null     object 
 4   Incoming or Outgoing   82 non-null     object 
 5   Queue                  82 non-null     object 
 6   Rep ID                 82 non-null     object 
 7   Sale                   82 non-null     object 
 8   Lost Call              82 non-null     int64  
 9   Abandoned              82 non-null     int64  
 10  Busy Minutes           82 non-null     int64  
 11  Not Ready Minutes      82 non-null     int64  
 12  Incoming Wait Time     82 non-null     float64
 13  During Call Wait Time  82 non-null     int64  
 14  Calls                  82 non-null     int64  
dtypes: float

#### Cleaning Question 2:  In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?
##### Record your answer by double clicking here:
> Your Answer: 
(1) By using the unique method for the column "Sale', I found a trailing space in the value of one variable from the output: array(['NO', 'YES ', 'YES'], dtype=object). 
(2) To remove the trailing space, the rstrip() method is used for the string.
array(['NO', 'YES'], dtype=object)

In [82]:
#Code Used
import pandas as pd
n2_df = pd.read_csv("NorthCallCenter.csv")
n2_df['Sale'].unique() # Initial check on distinct data values
n2_df['Sale'] = north_df['Sale'].str.rstrip()
n2_df['Sale'].unique() #  Validate only 'YES' and 'NO' existing in the column after space cleanup
# Add Additional Cells aorths needed.


array(['NO', 'YES'], dtype=object)

In [7]:
#### Cleaning Question 3:  Are there duplicates in the two data sets? If there are, how did you handle them?
##### Record your answer by double clicking here:
> Your Answer: With the use of drop_duplicates(), the dataframe will be left with unique rows.

SyntaxError: invalid syntax (<ipython-input-7-ae450db51b17>, line 3)

In [84]:
#Code Used
n2_df = n2_df.drop_duplicates()
n2_df.info()
s2_df = s2_df.drop_duplicates()
s2_df.info()
# Add Additional Cells as needed.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 245 entries, 0 to 244
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             245 non-null    int64  
 1   Branch                 245 non-null    object 
 2   Call Purpose           245 non-null    object 
 3   Time Block             245 non-null    object 
 4   Incoming or Outgoing   245 non-null    object 
 5   Queue                  245 non-null    object 
 6   Rep ID                 245 non-null    object 
 7   Sale                   245 non-null    object 
 8   Lost Call              245 non-null    int64  
 9   Abandoned              245 non-null    int64  
 10  Busy Minutes           245 non-null    int64  
 11  Not Ready Minutes      245 non-null    int64  
 12  Incoming Wait Time     82 non-null     float64
 13  During Call Wait Time  245 non-null    int64  
 14  Calls                  245 non-null    int64  
dtypes: flo

#### Cleaning Question 4:  Is any of the data in the two data sets unnecessary? If yes, how did you handle it?
##### Record your answer by double clicking here:
> Your Answer: Queue is of no use here and the column can be dropped using drop('column', axis=1, inplace=True).

n2_df.drop('Queue', axis=1, inplace=True)
Output:
Index(['Unnamed: 0', 'Branch', 'Call Purpose', 'Time Block',
       'Incoming or Outgoing', 'Rep ID', 'Sale', 'Lost Call', 'Abandoned',
       'Busy Minutes', 'Not Ready Minutes', 'Incoming Wait Time',
       'During Call Wait Time', 'Calls'],
      dtype='object')
s2_df.drop('Queue', axis=1, inplace=True)
Output:
Index(['Unnamed: 0', 'Branch', 'Call Purpose', 'Time Block',
       'Incoming or Outgoing', 'Rep ID', 'Sale', 'Lost Call', 'Abandoned',
       'Busy Minutes', 'Not Ready Minutes', 'Incoming Wait Time',
       'During Call Wait Time', 'Calls'],Abandoned
      dtype='object') 

In [102]:
#Code Used
#n2_df = pd.read_csv("NorthCallCenter.csv")
#n2_df['Sale'] = n2_df['Sale'].str.rstrip()
#n2_df['Sale'].unique()
#s2_df = pd.read_csv("SouthCallCenter.csv")

#  Drop the column "Queue" from North Branch dataframe
#n2_df.drop('Queue', axis=1, inplace=True)
#  Drop the column "Queue" from South Branch dataframe
#s2_df.drop('Queue', axis=1, inplace=True)
#n2_df.columns
#s2_df.columns
#n2_df.shape # Output: (245, 14)
s2_df.shape  # Output: (314. 14)
# Add Additional Cells as needed.

(314, 14)

## Data Manipulation
Before you begin answering any questions, combine the two datasets together to create a third dataframe called df. You can use this third dataframe to compare the two call centers to company-wide trends.
* Add additional code blocks as needed to show how you came to your conclusions. Add comments in your code blocks to help others understand your thinking. Record your answer by double clicking the below cell.

In [104]:
# Create dataframe for the entire company named df
df = n2_df.append(s2_df)
df.shape

(559, 14)

#### Manipulation Question 1:  Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.

##### Record your answer by double clicking here:
Rep with the hightest number of calls and their total calls:
- North Branch = Brent
- South Branch = Eric


In [108]:
# Code Used
df.groupby(['Rep ID','Branch'])[['Calls']].sum().sort_values(["Calls"], ascending=False)

# Add Additional Cells as needed.

Calls
Rep ID Branch       
Brent  North      37
Eric   South      35
Randy  South      33
Sandy  South      32
Helga  South      31
Lilly  North      30
Todd   North      29
Duke   North      29
George South      29
Kate   South      29
Karl   South      28
Josh   South      26
Susan  South      26
Joe    North      26
Sharon South      25
Xander North      24
Cam    North      24
Amanda North      23
Andy   North      23
Alice  South      20

#### Manipulation Question 2:  The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of thes. fall out of the optimal range of 2-5%?
##### Record your answer by double clicking here:
Average Call Abandonment Rates (round to the nearest hundredth):
- North Branch = 2.86%
- South Branch = 1.27%
- Company Wide = 1.97%
- Do any of these fall out of the optimal range of 2-5%? NO

In [129]:
# Code Used
# Calculate the average Abandoned rate for North Branch and South Branch 
print(df.groupby('Branch')[['Abandoned']].mean())

# Calculate the average Abandoned rate for the whole company
print("Company Wide: " , df[['Abandoned']].mean())

# Add Additional Cells as needed.

        Abandoned
Branch           
North    0.028571
South    0.012739
Company Wide:  Abandoned    0.019678
dtype: float64


#### Manipulation Question 3:  Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each branch and the entire company?
##### Record your answer by double clicking here:
Percentage of calls answered within 2 seconds, include # of calls:
- North Branch = 
- South Branch = 
- Company Wide = 

In [131]:
# Code Used
s2_df.info()
# Calculate the percentage of calls answered within 2 seconds
print(df.groupby('Branch')[['Incoming Wait Time']].mean())
#'Incoming Wait Time'
# Add Additional Cells as needed.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 314 entries, 0 to 313
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             314 non-null    int64  
 1   Branch                 314 non-null    object 
 2   Call Purpose           314 non-null    object 
 3   Time Block             314 non-null    object 
 4   Incoming or Outgoing   314 non-null    object 
 5   Rep ID                 314 non-null    object 
 6   Sale                   314 non-null    object 
 7   Lost Call              314 non-null    int64  
 8   Abandoned              314 non-null    int64  
 9   Busy Minutes           314 non-null    int64  
 10  Not Ready Minutes      314 non-null    int64  
 11  Incoming Wait Time     126 non-null    float64
 12  During Call Wait Time  314 non-null    int64  
 13  Calls                  314 non-null    int64  
dtypes: float64(1), int64(7), object(6)
memory usage: 34.5+ KB


#### Manipulation Question 4: For each branch and the entire company, what is the average speed of answer?
##### Record your answer by double clicking here:
Average speed of answer (rounded to nearest hundredth):
- North Branch in seconds = 
- South Branch in seconds = 
- Company Wide in seconds = 


In [ ]:
# Code Used 


<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Visualization

Create a visualization for each of the following questions. Some of the code to handle aggregating and storing data may be written for you. For each visualization, you choose the chart style that you feel suits the situation best. Make note of the chart style you chose and why. 

*NOTE Some questions you may decide to use more than one chart and or chart style.

#### Visualization 1:  What is the average abandonment rate per queue?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [ ]:
# Create visualization 1 here
# Import any additional libraries needed with alias



# The dictionary abandonment_rates has the data you need.
abandonment_rates = {}
queues = ["A", "B", "C", "D"]
queue_dict = df.groupby("Queue").agg("sum")
for i in range(4):
    abandonment_rates[queues[i]] = queue_dict["Abandoned"][i] / queue_dict["Calls"][i] 

#Your Code Here:




#### Visualization 2: What is the service level percentage and average speed of answer for each rep in the North Branch?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [ ]:
# Create visualization 2 here
# north_plt contains the data you need for the average speed of answer of each rep

north_plt = north.groupby("Rep ID")["Incoming Wait Time"].mean().to_frame().reset_index()

# Finding each Rep's Personal Service Level Percentage.  Basically, Calls within 2 secs / total calls

# Table 1: Total Incoming calls less than 2 seconds grouped by Rep
quick_calls = north[north["Incoming Wait Time"] <= 2.0]
quick_reps = quick_calls[["Rep ID", "Calls"]]
quick_stats = quick_reps.groupby(["Rep ID"]).sum()  # Final Table


# Table 2: Total Incoming Calls Only grouped by Rep
total_calls_in = north[north["Incoming or Outgoing"] == "Incoming"]
rep_calls = total_calls_in[["Rep ID", "Calls"]]     
total_stats = rep_calls.groupby(["Rep ID"]).sum() # Final Table  

#  Table 3: Service Level Percentage created via merge
service_level = pd.merge(quick_stats, total_stats, on="Rep ID")

# Create Percentage Column in Table 3
service_level["Service Level %"] = service_level["Calls_x"]/service_level["Calls_y"] * 100

#Your Code Here:


#### Visualization 3: For each type of call purpose, how many calls are outgoing vs. incoming?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [ ]:
# Create visualization 3 here:
# The three dictionaries, complaints, sales_support, and product_support, have the information you need

purpose_group = df.groupby("Call Purpose")
call_purpose = ["Complaint", "Product Support", "Sales Support"]
purpose_counts = purpose_group["Incoming or Outgoing"].value_counts()
print(purpose_counts)

complaints = purpose_counts["Complaint"].to_dict()
sales_support = purpose_counts["Sales Support"].to_dict()
product_support = purpose_counts["Product Support"].to_dict()

#Your Code Here:


<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Summarize Your Work

With what you know now about the two call centers and the entire company, answer the following questions. Note that while this is subjective, you should include relevant data to back up your opinion.


#### Summary Question 1:  Using KPIs such as average abandonment rate, service level and average speed of answer, in your opinion, which one of the two branches is operating more efficiently? Why?

##### Record your answer by double clicking here:
> Your Answer: 



#### Summary Question 2: Based on the number of reps in each branch and how quickly the reps are working, in your opinion, which branch would benefit from the extra help?

##### Record your answer by double clicking here:
> Your Answer: 


#### Summary Question 3: Now that you have explored the datasets, is there any data or information that you wish you had in this analysis?

##### Record your answer by double clicking here:
> Your Answer: 


<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Bonus Mission
Create a visualization that answers this question: For each call purpose, how many calls (incoming and outgoing) take place in each time block?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [ ]:
# Create your Bonus Mission visualization here!
call_times = df[["Time Block", "Call Purpose", "Incoming or Outgoing", "Calls"]]

# Use groupby to plot based on time blocks:

# Use groupby and get_group to select which call purpose to plot: